### PR TITLE
[SPARK-6077] Remove streaming tab while stopping StreamingContext

### PR DESCRIPTION
**Pull request by zhichao-li**
Currently we would create a new streaming tab for each streamingContext even if there's already one on the same sparkContext which would cause duplicate StreamingTab created and none of them is taking effect.
snapshot: https://www.dropbox.com/s/t4gd6hqyqo0nivz/bad%20multiple%20streamings.png?dl=0
How to reproduce:
1)
import org.apache.spark.SparkConf
import org.apache.spark.streaming.
{Seconds, StreamingContext}
import org.apache.spark.storage.StorageLevel
val ssc = new StreamingContext(sc, Seconds(1))
val lines = ssc.socketTextStream("localhost", 9999, StorageLevel.MEMORY_AND_DISK_SER)
val words = lines.flatMap(_.split(" "))
val wordCounts = words.map(x => (x, 1)).reduceByKey(_ + _)
wordCounts.print()
ssc.start()
.....
2)
ssc.stop(false)
val ssc = new StreamingContext(sc, Seconds(1))
val lines = ssc.socketTextStream("localhost", 9999, StorageLevel.MEMORY_AND_DISK_SER)
val words = lines.flatMap(_.split(" "))
val wordCounts = words.map(x => (x, 1)).reduceByKey(_ + _)
wordCounts.print()
ssc.start()

### CODE REVIEW

File: streaming/src/main/scala/org/apache/spark/streaming/ui/StreamingTab.scala [hunk 1]
```diff
@@ -29,19 +29,30 @@ import StreamingTab._
...
 }
 
-private object StreamingTab {
+private[spark] object StreamingTab {
   def getSparkUI(ssc: StreamingContext): SparkUI = {
     ssc.sc.ui.getOrElse {
       throw new SparkException("Parent SparkUI to attach this tab to not found!")
     }
   }
+  def updateOrCreateStreamingTab(ssc:StreamingContext):Option[StreamingTab]={
```
Comment (srowen) [inline streaming/src/main/scala/org/apache/spark/streaming/ui/StreamingTab.scala]: You have some style issues here, like spaces around operators, braces, and parentheses. Have a look at other code. Also we usually write e.g. `foreach { tab =>` not `foreach(tab = {`
I think the code here can be more concise and avoid `return` with a construction built around `getTabs.find(_.isInstanceOf[StreamingTab]).map(...).getOrElse(...)

Comment (zhichao-li) [inline streaming/src/main/scala/org/apache/spark/streaming/ui/StreamingTab.scala]: @srowen , Thanks for the advice. I've just tested and refactored the code accordingly.

File: streaming/src/main/scala/org/apache/spark/streaming/ui/StreamingPage.scala [hunk 2]
```diff
@@ -20,20 +20,26 @@ package org.apache.spark.streaming.ui
...
   extends WebUIPage("") with Logging {
 
-  private val listener = parent.listener
-  private val startTime = Calendar.getInstance().getTime()
+  private var listener = parent.listener
+  private var startTime = Calendar.getInstance().getTime()
   private val emptyCell = "-"
+  
+  def updateListener(listener : StreamingJobProgressListener):
+  Unit ={
```
Comment (srowen) [inline streaming/src/main/scala/org/apache/spark/streaming/ui/StreamingPage.scala]: The wrapping is wrong here; `Unit` should be on the previous line and no space before the brace. No space before `:`. Also, `System.currentTimeMillis()` is much more conventional for getting the time.

File: streaming/src/main/scala/org/apache/spark/streaming/ui/StreamingTab.scala [hunk 3]
```diff
@@ -29,19 +29,29 @@ import StreamingTab._
...
 
-private object StreamingTab {
+private[spark] object StreamingTab {
   def getSparkUI(ssc: StreamingContext): SparkUI = {
     ssc.sc.ui.getOrElse {
       throw new SparkException("Parent SparkUI to attach this tab to not found!")
     }
   }
+  
+  def updateOrCreateStreamingTab(ssc: StreamingContext): Option[StreamingTab] = {
```
Comment (srowen) [inline streaming/src/main/scala/org/apache/spark/streaming/ui/StreamingTab.scala]: This method does not need to return `Option`. It always returns `StreamingTab`.

File: streaming/src/main/scala/org/apache/spark/streaming/ui/StreamingPage.scala (hunk 2 above)
Comment (zhichao-li) [inline streaming/src/main/scala/org/apache/spark/streaming/ui/StreamingPage.scala]: Done.

File: streaming/src/main/scala/org/apache/spark/streaming/ui/StreamingTab.scala (hunk 3 above)
Comment (zhichao-li) [inline streaming/src/main/scala/org/apache/spark/streaming/ui/StreamingTab.scala]: Updated accordingly.

**Comment by srowen** [comment]
This looks OK to me. I'd like to give @tdas at least a day to look at it.

**Comment by tdas** [comment]
Hey this is a decent fix, but I think this is not the right one. With this fix, the new ssc will be reflected in the new streaming tab, however it will be still visible even after the earlier ssc has been stopped. The right solution is that the tab should be removed when a stream ing context is stopped. Since only one streaming context can be active on the same spark context at the same time, attach-on-start-and-remove-on-stop will fix the multiple tab problem in the right way. 

Does that make sense?

**Comment by JoshRosen** [comment]
@tdas do you think we should add a Selenium test for this?

**Comment by tdas** [comment]
Will that be stable? Not flaky? In the past we had simple web ui tests that
uses Scala's Source class to fetch a URL to see whether a tab has been
loaded or unloaded. Those were disabled because of flakiness. I wonder
whether selenium tests will be more stable.

**Comment by JoshRosen** [comment]
Poorly-written Selenium tests can be flaky if they don't account for things like asynchrony (e.g. when testing Javascript interactions), but I don't think that will be a problem here.  We now have a bunch of Selenium tests for the Spark core UI and I don't think I've ever seen them fail in Jenkins: https://github.com/apache/spark/blob/master/core/src/test/scala/org/apache/spark/ui/UISeleniumSuite.scala

**Comment by tdas** [comment]
Then we should go for it!

**Comment by zhichao-li** [comment]
@tdas  Just updated the code with the solution you mentioned(attach-on-start-and-remove-on-stop) by adding the detachTab and detachPage function.

File: core/src/main/scala/org/apache/spark/ui/WebUI.scala [hunk 4]
```diff
@@ -17,17 +17,17 @@
 
 package org.apache.spark.ui
 
+import scala.collection.mutable.HashMap
```
Comment (jerryshao) [inline core/src/main/scala/org/apache/spark/ui/WebUI.scala]: Change the import ordering.

```
Java import

Scala import 

Third party import 

Spark import 
```

File: core/src/main/scala/org/apache/spark/ui/WebUI.scala [hunk 5]
```diff
@@ -60,14 +61,34 @@ private[spark] abstract class WebUI(
...
     tabs += tab
   }
+  
+  def detachTab(tab: WebUITab) {
+    tab.pages.foreach(detachPage)
+    tabs -= tab
+  }
+  
+  def detachPage(page: WebUIPage) {
+    page2Handlers.remove(page) match {
```
Comment (jerryshao) [inline core/src/main/scala/org/apache/spark/ui/WebUI.scala]: This can be simplified as `page2Handles.remove(page).foreach(...)`

File: streaming/src/main/scala/org/apache/spark/streaming/StreamingContext.scala [hunk 6]
```diff
@@ -553,6 +553,10 @@ class StreamingContext private[streaming] (
    */
   def stop(stopSparkContext: Boolean = true): Unit = synchronized {
     stop(stopSparkContext, false)
+    this.uiTab match  {
```
Comment (jerryshao) [inline streaming/src/main/scala/org/apache/spark/streaming/StreamingContext.scala]: The same as above.

Comment (zhichao-li) [inline streaming/src/main/scala/org/apache/spark/streaming/StreamingContext.scala]: Thanks @jerryshao . Done with the changes.

File: core/src/main/scala/org/apache/spark/ui/WebUI.scala [hunk 7]
```diff
@@ -45,6 +46,7 @@ private[spark] abstract class WebUI(
 
   protected val tabs = ArrayBuffer[WebUITab]()
   protected val handlers = ArrayBuffer[ServletContextHandler]()
+  protected val page2Handlers = new HashMap[WebUIPage, ArrayBuffer[ServletContextHandler]]
```
Comment (tdas) [inline core/src/main/scala/org/apache/spark/ui/WebUI.scala]: We dont use the idiom "to --> 2" in our code. `pageToHandlers` is better.

File: core/src/main/scala/org/apache/spark/ui/WebUI.scala [hunk 8]
```diff
@@ -60,14 +62,31 @@ private[spark] abstract class WebUI(
...
+    tabs -= tab
+  }
+  
+  def detachPage(page: WebUIPage) {
+    page2Handlers.remove(page).foreach(_.foreach(detachHandler))
+  }
+  
 
   /** Attach a page to this UI. */
   def attachPage(page: WebUIPage) {
```
Comment (tdas) [inline core/src/main/scala/org/apache/spark/ui/WebUI.scala]: @JoshRosen Can you take a look at this code. This is your domain.

File: streaming/src/main/scala/org/apache/spark/streaming/StreamingContext.scala [hunk 9]
```diff
@@ -553,6 +553,7 @@ class StreamingContext private[streaming] (
    */
   def stop(stopSparkContext: Boolean = true): Unit = synchronized {
     stop(stopSparkContext, false)
+    this.uiTab.foreach(StreamingTab.detachStreamingTab(this, _))
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/StreamingContext.scala]: Why is it hear and not in the other `stop` The tab should be removed when the stremaing context is stopped, so both `stop` should detach the tab.

File: streaming/src/main/scala/org/apache/spark/streaming/ui/StreamingTab.scala [hunk 10]
```diff
@@ -38,10 +38,13 @@ private[spark] class StreamingTab(ssc: StreamingContext)
...
 }
 
-private object StreamingTab {
+private[spark] object StreamingTab {
   def getSparkUI(ssc: StreamingContext): SparkUI = {
     ssc.sc.ui.getOrElse {
       throw new SparkException("Parent SparkUI to attach this tab to not found!")
     }
   }
+  def detachStreamingTab(ssc: StreamingContext, tab: SparkUITab) {
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/ui/StreamingTab.scala]: Need extra line.

**Comment by tdas** [comment]
Can you add unit tests for detaching pages and tabs. @JoshRosen will be able to help you out.

**Comment by tdas** [comment]
BTW, please update the title and JIRA as we are not updating existing tab any more.

**Comment by JoshRosen** [comment]
Spark core's `UISuite` has an ignored test for "attaching a new tab": https://github.com/apache/spark/blob/dba0b2eadb441f41ded0f0b9706b720bcfa84881/core/src/test/scala/org/apache/spark/ui/UISuite.scala#L75

In order to test the tab attach / detach functionality, I think we should re-enable and extend this test to check that we're able to remove the tab that we just attached.  It looks like this test may have been disabled due to port contention issues (#466), but I think those issues should be resolved now.  If it will be easier, you might rewriting this test using the new [UISeleniumSuite](https://github.com/apache/spark/blob/dba0b2eadb441f41ded0f0b9706b720bcfa84881/core/src/test/scala/org/apache/spark/ui/UISeleniumSuite.scala) framework, which will let you use Selenium CSS selectors, etc. in your test.

**Comment by zhichao-li** [comment]
@JoshRosen , the test for tab attach / detach has been added base on the new UISeleniumSuite framework. Could you help to review?

**Comment by JoshRosen** [comment]
The new attach / detach test looks fine to me.

Jenkins, this is ok to test.

**Comment by tdas** [comment]
ok to test

File: streaming/src/main/scala/org/apache/spark/streaming/ui/StreamingTab.scala [hunk 11]
```diff
@@ -38,10 +38,14 @@ private[spark] class StreamingTab(ssc: StreamingContext)
...
 }
 
-private object StreamingTab {
+private[spark] object StreamingTab {
   def getSparkUI(ssc: StreamingContext): SparkUI = {
     ssc.sc.ui.getOrElse {
       throw new SparkException("Parent SparkUI to attach this tab to not found!")
     }
   }
+
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/ui/StreamingTab.scala]: Can you move this code into the StreamingTab class? The code is related to the tab instance, and so should be in the class. You could argue that `getSparkUI` should also be in the class, however it needs to be called from constructor (line 31) and hence cannot be in the class. But the `detach` function should be in the class.

File: streaming/src/main/scala/org/apache/spark/streaming/StreamingContext.scala [hunk 12]
```diff
@@ -578,6 +578,7 @@ class StreamingContext private[streaming] (
     // Even if we have already stopped, we still need to attempt to stop the SparkContext because
     // a user might stop(stopSparkContext = false) and then call stop(stopSparkContext = true).
     if (stopSparkContext) sc.stop()
+    uiTab.foreach(StreamingTab.detachStreamingTab(this, _))
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/StreamingContext.scala]: If the detach is moved, this should become `uiTab.foreach { _.detach() }`

Comment (zhichao-li) [inline streaming/src/main/scala/org/apache/spark/streaming/StreamingContext.scala]: @tdas the code has been updated. Thanks for your review.

**Comment by zhichao-li** [comment]
Jenkins, retest this please.

**Comment by tdas** [comment]
add to whitelist

**Comment by tdas** [comment]
ok to test

**Comment by tdas** [comment]
LGTM, assuming the tests pass.

**Comment by tdas** [comment]
Oh wait. I just realized that there is a already a unit test in org.apace.spark.streaming.UISuite which tested whether the streaming tab works. But thats ignored too. Could you update that testsuite to use selenium and test whether the streaming tab is attached on and detached correctly. I just realized that the unit tests in the PR tests detaching for tabs, but does not test whether it works specifically for streaming (suppose the line to detach gets deleted accidentally in future, no unit test would fail). So please add the streaming's specific unit test - rename spark.streaming.UISuite to UISeleniumSuite, and then update test to use selenium.

**Comment by zhichao-li** [comment]
@tdas UISeleniumSuite has been added for streaming and the LocalStreamingContext utilities is borrowed from LocalSparkContext .

File: streaming/src/test/scala/org/apache/spark/streaming/LocalStreamingContext.scala [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+  def stop(ssc: StreamingContext) {
+    if (ssc != null) {
+      ssc.stop()
+    }
+    // To avoid Akka rebinding to the same port, since it doesn't unbind immediately on shutdown
+    System.clearProperty("spark.driver.port")
+  }
+
+  /** Runs `f` by passing in `ssc` and ensures that `ssc` is stopped. */
+  def withSparkStreaming[T](ssc: StreamingContext)(f: StreamingContext => T) = {
```
Comment (tdas) [inline streaming/src/test/scala/org/apache/spark/streaming/LocalStreamingContext.scala]: This is already defined in streaming.TestSuiteBase. Rather than creating this new file, please use that.

File: streaming/src/test/scala/org/apache/spark/streaming/UISeleniumSuite.scala [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+    ssc
+  }
+
+  test("attaching and detaching a Streaming tab") {
+    withSparkStreaming(newSparkStreamingContext()) { ssc =>
+      val sparkUI = ssc.sparkContext.ui.get
+
+      eventually(timeout(10 seconds), interval(50 milliseconds)) {
+        go to (sparkUI.appUIAddress.stripSuffix("/"))
+        find(cssSelector( """ul li a[href*="jobs"]""")) should not be (None)
```
Comment (tdas) [inline streaming/src/test/scala/org/apache/spark/streaming/UISeleniumSuite.scala]: I dont think the usual ones like jobs, stages, etc need to be tested. Just testing streaming tab should be fine. The test will be faster i hope.

Comment (zhichao-li) [inline streaming/src/test/scala/org/apache/spark/streaming/UISeleniumSuite.scala]: @tdas done.

**Comment by tdas** [comment]
LGTM! Will merge asap.

**Comment by tdas** [comment]
Merged this in master and branch-1.3
